lib/cadet/assessments/assessments.ex

```diff
@@ -371,6 +372,73 @@ defmodule Cadet.Assessments do
     end
   end
 
+  def unsubmit_submission(submission_id, avenger = %User{role: role})
+      when is_ecto_id(submission_id) do
+    if role in @unsubmit_assessment_role do
+      submission =
+        Submission
+        |> join(:inner, [s], a in assoc(s, :assessment))
+        |> preload([_, a], assessment: a)
+        |> Repo.get(submission_id)
+
+      with {:submission_found?, true} <- {:submission_found?, is_map(submission)},
+           {:is_open?, true} <- is_open?(submission.assessment),
+           {:status, :submitted} <- {:status, submission.status},
+           {:avenger_of?, true} <-
+             {:avenger_of?, Cadet.Accounts.Query.is_avenger_of(avenger, submission.student_id)} do
+        Ecto.Multi.new()
+        |> Multi.run(
+          :rollback_submission,
+          fn _ ->
+            submission
+            |> Submission.changeset(%{status: :attempted, xp_bonus: 0})
+            |> Repo.update()
+          end
+        )
+        |> Multi.run(:rollback_answers, fn _ ->
+          Answer
+          |> where(submission_id: ^submission.id)
+          |> Repo.all()
+          |> Enum.each(fn answer ->
+            answer
+            |> Answer.reset(%{
+              grade: 0,
+              adjustment: 0,
+              xp: 0,
+              xp_adjustment: 0,
+              autograding_status: :none,
+              autograding_errors: [],
+              comment: nil
+            })
+            |> Repo.update()
+          end)
+
+          {:ok, nil}
+        end)
+        |> Repo.transaction()
+
+        {:ok, nil}
+      else
+        {:submission_found?, false} ->
+          {:error, {:not_found, "Submission not found"}}
+
+        {:is_open?, false} ->
+          {:error, {:forbidden, "Assessment not open"}}
+
+        {:status, :attempting} ->
+          {:error, {:bad_request, "Some questions have not been attempted"}}
+
+        {:status, :attempted} ->
+          {:error, {:forbidden, "Assessment has not been submitted"}}
+
+        _ ->
```

Comment:
Let it crash! I don't think there's a need for this clause. Also you're missing the clause for `{:avenger_of?, false}`